gcloud-java-contrib/gcloud-java-nio/src/main/java/com/google/gcloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -532,13 +563,23 @@ public void createDirectory(Path dir, FileAttribute<?>... attrs) {
     checkNotNullArray(attrs);
   }
 
-  /**
-   * Throws {@link UnsupportedOperationException} because this feature hasn't been implemented yet.
-   */
   @Override
   public DirectoryStream<Path> newDirectoryStream(Path dir, Filter<? super Path> filter) {
-    // TODO: Implement me.
-    throw new UnsupportedOperationException();
+    CloudStoragePath cloudPath = checkPath(dir);
+    checkNotNull(filter);
+    String prefix = cloudPath.toString();
+    Iterator<Blob> blobIterator = storage.list(cloudPath.bucket(), Storage.BlobListOption.prefix(prefix), Storage.BlobListOption.fields()).iterateAll();
+    final Iterator<Path> pathIterator = new LazyPathIterator(cloudPath.getFileSystem(), blobIterator, filter);
+    return new DirectoryStream<Path>() {
+      @Override
+      public Iterator<Path> iterator() {
+        return pathIterator;
+      }
+
+      @Override
+      public void close() throws IOException {
+      }
```

Comment:
Even though the comment here is not really necessary it might be useful to show that we did not forget and to silent codacy.